DR: Define who should validate a decision recording PR (#31)
blindnet-io/openness-framework: docs/decision-records/DR-0002-dr-validation.md

# Decision Recording Pull Request Validation Protocol

| Status      | proposed                                            |
| :---------- | :-------------------------------------------------- |
| **PR #**    | [31](https://github.com/blindnet-io/openness-framework/pull/31) |
| **Author** | milstan (milstan@blindnet.io)                        |

## Context and Problem Statement

> As blindnet [decided](https://github.com/blindnet-io/devrel-planning/issues/49) to fully embrace Openness principles, and has put in place a [Decision Framework](https://github.com/blindnet-io/openness-framework/tree/main/DecisionFramework), the need has emerged to establish more clarity in the process of proposing and approving Decision Records.

Reviewer: ```suggestion
As blindnet [decided](https://github.com/blindnet-io/devrel-management/issues/32) to fully embrace Openness principles, and has put in place a [Decision Framework](https://github.com/blindnet-io/openness-framework/tree/main/DecisionFramework), the need has emerged to establish more clarity in the process of proposing and approving Decision Records.
```

## Terminology

- The key words "MUST", "MUST NOT", "REQUIRED", "SHALL", "SHALL NOT", "SHOULD", "SHOULD NOT", "RECOMMENDED", "MAY", and "OPTIONAL" in this document are to be interpreted as described in [RFC 2119 of the IETF](https://datatracker.ietf.org/doc/html/rfc2119)
- We use the terms **decision** and **choice** as defined in the [Decision Framework](https://github.com/blindnet-io/openness-framework/tree/main/DecisionFramework)
- We use the term **Decision Recording**, to refer to recording of any **decision** regardless of the type of document used to record it (such as : DRs (_Decision Records_), SGs (_SMART Goals_) and RFCs (_Request For Comments_)).
- The terms "Assignee", "Collaborator", "Contributor", "Pull Request" (same as "PR"), and "Review" are to be interpreted as described in the [Github Glossary](https://docs.github.com/en/get-started/quickstart/github-glossary)
- A Reviewer is a Collaborator asked to do a Review of a Pull Request

> ## Decision Drivers <!-- optional -->

Reviewer: ```suggestion
## Decision Drivers
```

> ## Decision Drivers <!-- optional -->

Reviewer: ```suggestion
## Decision Drivers
```

* High clarity for Collaborators and Contributors about who should do what in order for a **Decision Recording** Pull Request to be validated
* Avoid getting the **Decision Recording** Pull Request validation process in a stale point

## Considered Options

Different options have been considered in the discussion that took place in [issue #14](https://github.com/blindnet-io/openness-framework/issues/14). As a result, the following two protocols are defined.

## Decision Outcome

### Validation protocol for template-based decisions

> >This protocol applies when recording **decisions** that follow the Decision Framework and use one of [its specific document templates](https://github.com/blindnet-io/openness-framework/tree/main/DecisionFramework#what-template-will-you-use).

Reviewer: ```suggestion
> This protocol applies when recording **decisions** that follow the Decision Framework and use one of [its specific document templates](https://github.com/blindnet-io/openness-framework/tree/main/DecisionFramework#what-template-will-you-use).
```

> >This protocol applies when recording **decisions** that follow the Decision Framework and use one of [its specific document templates](https://github.com/blindnet-io/openness-framework/tree/main/DecisionFramework#what-template-will-you-use).

Reviewer: ```suggestion
> This protocol applies when recording **decisions** that follow the Decision Framework and use one of [its specific document templates](https://github.com/blindnet-io/openness-framework/tree/main/DecisionFramework#what-template-will-you-use).
```

Anyone MAY, and at least one of the {**Corresponding Author**, **Sponsor**} MUST designate one or more eligible Reviewers.
An eligible Reviewer MAY designate themselves.

The **Corresponding Author** OR the **Sponsor** MUST assign the **Decision Recording** Pull Request to an eligible Assignee.
The Assignee MAY also be a Reviewer.

The Assignee looks for consensus before merging the **Decision Recording** Pull Request.
In clear absence of consensus the Assignee closes the request.
It is the responsibility of the Assignee to act and not let PR remain at a stale point.

Whenever possible (unless they are not eligible, or for other reasons known to be unavailable), the Assignee SHOULD be:

1. `@noelmace` for everything related to communication, openness framework, DevX and DevRel in general
2. `@filipblnt` for everything related to engineering, by default (adding a member of the engineering team as an additional Reviewer whenever needed)
3. `@milstan` for everything related to our company's "vision" and high-level concepts (i.e. high-level product ownership)
4. `@Vuk-BN` for everything related to global executive strategies and finances

> When such assignment is not possible, the then the Assignee SHOULD be the first eligible and available person from the above list, in the order from 1. to 4.

Reviewer: ```suggestion
When such an assignment is not possible, then the Assignee SHOULD be the first eligible and available person from the above list, in the order from 1. to 4.
```

> When such assignment is not possible, the then the Assignee SHOULD be the first eligible and available person from the above list, in the order from 1. to 4.

Reviewer: ```suggestion
When such an assignment is not possible, then the Assignee SHOULD be the first eligible and available person from the above list, in the order from 1. to 4.
```

### Validation protocol for non template-based documents

> >This protocol applies, unless a particular project-related workflow is specified in the project's readme file, when producing other types of documents that MAY also involve more or less normative **choices** and even certain **decisions** (e.g. user stories, specifications, designs, documentation work, code) but that do not use one of the [Decision Framework's specific document templates](https://github.com/blindnet-io/openness-framework/tree/main/DecisionFramework#what-template-will-you-use)

Reviewer: ```suggestion
> This protocol applies, unless a particular project-related workflow is specified in the project's readme file, when producing other types of documents that MAY also involve more or less normative **choices** and even certain **decisions** (e.g. user stories, specifications, designs, documentation work, code) but that do not use one of the [Decision Framework's specific document templates](https://github.com/blindnet-io/openness-framework/tree/main/DecisionFramework#what-template-will-you-use)
```

> >This protocol applies, unless a particular project-related workflow is specified in the project's readme file, when producing other types of documents that MAY also involve more or less normative **choices** and even certain **decisions** (e.g. user stories, specifications, designs, documentation work, code) but that do not use one of the [Decision Framework's specific document templates](https://github.com/blindnet-io/openness-framework/tree/main/DecisionFramework#what-template-will-you-use)

Reviewer: ```suggestion
> This protocol applies, unless a particular project-related workflow is specified in the project's readme file, when producing other types of documents that MAY also involve more or less normative **choices** and even certain **decisions** (e.g. user stories, specifications, designs, documentation work, code) but that do not use one of the [Decision Framework's specific document templates](https://github.com/blindnet-io/openness-framework/tree/main/DecisionFramework#what-template-will-you-use)
```

One of the Authors, the **Corresponding Author**, makes a PR.

> They invite Reviewers others then themselves.

Reviewer: ```suggestion
They invite Reviewers other than themselves.
```

> They invite Reviewers others then themselves.

Reviewer: ```suggestion
They invite Reviewers other than themselves.
```

> A Reviewer MAY designate themselves.

Reviewer: Reading this, I think it could be unclear _to what_ and _how_ a rewiewer may designate themselves.

The following might be a better phrasing:
```suggestion
Anyone MAY [review a PR](https://github.com/blindnet-io/openness-framework/blob/main/docs/HOWTOs/PR-review.md), therefore designating themselves as a **Reviewer**.
```

> A Reviewer MAY designate themselves.

Reviewer: Reading this, I think it could be unclear _to what_ and _how_ a rewiewer may designate themselves.

The following might be a better phrasing:
```suggestion
Anyone MAY [review a PR](https://github.com/blindnet-io/openness-framework/blob/main/docs/HOWTOs/PR-review.md), therefore designating themselves as a **Reviewer**.
```

The **Corresponding Author** assigns a person (other than themselves, which MAY also be a Reviewer, in the same order as for template-based decisions).

The Assignee looks for consensus before merging the PR.

> In clear absence of consensus the Assignee closes the request.

Reviewer: I'm not sure about this one (and maybe even the same point in the previous section).

In some cases, the assignee might still need to make a choice when some reviewers _slightly_ disagree. We might need to add a definition for a **formal objection** here (as opposed to an informal objection, when a reviewer _slightly_ disagrees, but not enough to block the decision). As always, I take my inspiration from Open Standards Organizations (here, [W3C's Processing of Formal Objections](https://www.w3.org/2017/12/formal-objections.html)):

> **Best efforts to resolve via consensus.**
> 
> The delegate should search for potential consensus positions. When that is not possible, at least the delegate is in a position to clearly articulate the rationale for the decision.
> 
> **Analysis of all arguments raised.**
> 
> Although the decision should be **clear, well explained, and unambiguous**, the decision should nonetheless acknowledge good points raised by the party that is not favoured in the ruling, but explain why they are not decisive. 

However, here, closing a PR doesn't necessarily mean ending the decision process, as the discussion might continue in another issue or PR. So we might need to add a definition _"alternative proposal"_ instead or additionnaly.

Author: I propose tho address this by including a sepcial key word for Formal Objections, and by including it also in the HOW TO Review. See last commit.

Reviewer: Love this :100:

> In clear absence of consensus the Assignee closes the request.

Reviewer: I'm not sure about this one (and maybe even the same point in the previous section).

In some cases, the assignee might still need to make a choice when some reviewers _slightly_ disagree. We might need to add a definition for a **formal objection** here (as opposed to an informal objection, when a reviewer _slightly_ disagrees, but not enough to block the decision). As always, I take my inspiration from Open Standards Organizations (here, [W3C's Processing of Formal Objections](https://www.w3.org/2017/12/formal-objections.html)):

> **Best efforts to resolve via consensus.**
> 
> The delegate should search for potential consensus positions. When that is not possible, at least the delegate is in a position to clearly articulate the rationale for the decision.
> 
> **Analysis of all arguments raised.**
> 
> Although the decision should be **clear, well explained, and unambiguous**, the decision should nonetheless acknowledge good points raised by the party that is not favoured in the ruling, but explain why they are not decisive. 

However, here, closing a PR doesn't necessarily mean ending the decision process, as the discussion might continue in another issue or PR. So we might need to add a definition _"alternative proposal"_ instead or additionnaly.
